File: plugin-core/src/main/java/com/github/catatafishen/agentbridge/settings/ClaudeCliClientConfigurable.java
package com.github.catatafishen.agentbridge.settings;

import com.github.catatafishen.agentbridge.acp.client.AcpClient;
import com.github.catatafishen.agentbridge.services.AgentProfile;
import com.github.catatafishen.agentbridge.services.AgentProfileManager;
import com.github.catatafishen.agentbridge.ui.ThemeColor;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.options.Configurable;
import com.intellij.openapi.project.Project;
import com.intellij.ui.components.JBLabel;
import com.intellij.ui.components.JBScrollPane;
import com.intellij.ui.components.JBTextArea;
import com.intellij.ui.components.JBTextField;
import com.intellij.util.ui.FormBuilder;
import com.intellij.util.ui.JBUI;
import com.intellij.util.ui.UIUtil;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.awt.*;
import java.util.Arrays;
import java.util.List;

public final class ClaudeCliClientConfigurable implements Configurable {

    /**
     * CSS client type shared with Claude Code (AnthropicDirect) for bubble color.
     */
    private static final String BUBBLE_CLIENT_TYPE = "claude";

    @SuppressWarnings("unused")
    public ClaudeCliClientConfigurable(@NotNull Project ignoredProject) {
    }

    @Override
    public @Nls(capitalization = Nls.Capitalization.Title) String getDisplayName() {
        return "Claude CLI";
    }

    private JBLabel authStatusLabel;
    private JBTextField binaryPathField;
    private JBTextField instructionsFileField;
    private JBTextArea customModelsArea;
    private @Nullable ThemeColorComboBox bubbleColorCombo;
    private JPanel panel;

    @Override
    public @NotNull JComponent createComponent() {
        authStatusLabel = new JBLabel();

        binaryPathField = new JBTextField();
        binaryPathField.getEmptyText().setText("Auto-detect (leave empty)");
        binaryPathField.setToolTipText("Absolute path to the claude binary. Leave empty to find it on PATH.");

        instructionsFileField = new JBTextField();
        instructionsFileField.getEmptyText().setText("E.g. CLAUDE.md");
        instructionsFileField.setToolTipText(
            "Path relative to project root. Plugin instructions are prepended to this file on each session start.");

        customModelsArea = new JBTextArea(4, 40);
        customModelsArea.setLineWrap(false);
        customModelsArea.setFont(new Font(Font.MONOSPACED, Font.PLAIN, JBUI.Fonts.label().getSize()));
        JBScrollPane modelsScroll = new JBScrollPane(customModelsArea);
        modelsScroll.setPreferredSize(JBUI.size(400, 90));

        bubbleColorCombo = new ThemeColorComboBox();

        JBLabel authNote = new JBLabel(
            "<html>Run <code>claude auth login</code> in a terminal to authenticate.</html>");
        authNote.setForeground(UIUtil.getContextHelpForeground());
        authNote.setFont(JBUI.Fonts.smallFont());

        panel = FormBuilder.createFormBuilder()
            .addLabeledComponent("Status:", authStatusLabel)
            .addComponent(authNote, 2)
            .addSeparator(8)
            .addLabeledComponent("Claude binary:", binaryPathField)
            .addTooltip("Leave empty to auto-detect on PATH.")
            .addLabeledComponent("Instructions file:", instructionsFileField)
            .addTooltip("Plugin instructions are prepended here on session start (relative to project root).")
            .addLabeledComponent("Bubble color:", bubbleColorCombo)
            .addTooltip("Choose a theme-aware accent color for Claude message bubbles. Shared with Claude Code.")
            .addSeparator(8)
            .addComponent(new JBLabel("Custom models (one per line):"))
            .addTooltip("Format: <model-id>=<Display Name>. Leave empty to use the built-in model list.")
            .addComponentFillVertically(modelsScroll, 0)
            .getPanel();
        panel.setBorder(JBUI.Borders.empty(8));
        reset();
        return panel;
    }

    @Override
    public boolean isModified() {
        if (customModelsArea == null) return false;
        AgentProfile p = AgentProfileManager.getInstance().getProfile(AgentProfileManager.CLAUDE_CLI_PROFILE_ID);
        if (p == null) return false;
        if (!parseModels().equals(p.getCustomCliModels())) return true;
        if (!binaryPathField.getText().trim().equals(p.getCustomBinaryPath())) return true;
        if (!instructionsFileField.getText().trim().equals(nullToEmpty(p.getPrependInstructionsTo()))) return true;
        if (bubbleColorCombo != null) {
            ThemeColor tc = bubbleColorCombo.getSelectedThemeColor();
            String key = tc != null ? tc.name() : null;
            if (!java.util.Objects.equals(key, AcpClient.loadAgentBubbleColorKey(BUBBLE_CLIENT_TYPE))) return true;
        }
        return false;
    }

    @Override
    public void apply() {
        if (customModelsArea == null) return;
        AgentProfileManager mgr = AgentProfileManager.getInstance();
        AgentProfile p = mgr.getProfile(AgentProfileManager.CLAUDE_CLI_PROFILE_ID);
        if (p == null) return;
        p.setCustomCliModels(parseModels());
        p.setCustomBinaryPath(binaryPathField.getText().trim());
        p.setPrependInstructionsTo(instructionsFileField.getText().trim());
        if (bubbleColorCombo != null) {
            ThemeColor tc = bubbleColorCombo.getSelectedThemeColor();
            AcpClient.saveAgentBubbleColorKey(BUBBLE_CLIENT_TYPE, tc != null ? tc.name() : null);
        }
    }

    @Override
    public void reset() {
        if (customModelsArea == null) return;
        refreshAuthStatusAsync();
        AgentProfile p = AgentProfileManager.getInstance().getProfile(AgentProfileManager.CLAUDE_CLI_PROFILE_ID);
        if (p == null) return;
        binaryPathField.setText(p.getCustomBinaryPath());
        instructionsFileField.setText(nullToEmpty(p.getPrependInstructionsTo()));
        customModelsArea.setText(String.join("\n", p.getCustomCliModels()));
        customModelsArea.setCaretPosition(0);
        if (bubbleColorCombo != null) {
            bubbleColorCombo.setSelectedThemeColor(ThemeColor.fromKey(AcpClient.loadAgentBubbleColorKey(BUBBLE_CLIENT_TYPE)));
        }
    }

    @Override
    public void disposeUIResources() {
        authStatusLabel = null;
        binaryPathField = null;
        instructionsFileField = null;
        customModelsArea = null;
        bubbleColorCombo = null;
        panel = null;
    }

    private void refreshAuthStatusAsync() {
        if (authStatusLabel == null) return;
        authStatusLabel.setText("Checking...");
        authStatusLabel.setForeground(UIUtil.getLabelForeground());

        ApplicationManager.getApplication().executeOnPooledThread(() -> {
            String status = AgentProfileManager.getClaudeCliAuthStatus();
            SwingUtilities.invokeLater(() -> {
                if (authStatusLabel == null) return;
                if (status != null) {
                    authStatusLabel.setText("✓ Logged in — " + status);
                    authStatusLabel.setForeground(new Color(0, 128, 0));
                } else {
                    authStatusLabel.setText("Not logged in — run 'claude auth login' in a terminal");
                    authStatusLabel.setForeground(Color.RED);
                }
            });
        });
    }

    @NotNull
    private List<String> parseModels() {
        if (customModelsArea == null) return List.of();
        return Arrays.stream(customModelsArea.getText().split("\n"))
            .map(String::trim)
            .filter(s -> !s.isEmpty())
            .toList();
    }

    @NotNull
    private static String nullToEmpty(@Nullable String s) {
        return s != null ? s : "";
    }
}

File: plugin-core/src/main/java/com/github/catatafishen/agentbridge/settings/ClaudeCliClientConfigurable.kt
package com.github.catatafishen.agentbridge.settings

import com.github.catatafishen.agentbridge.acp.client.AcpClient
import com.github.catatafishen.agentbridge.services.AgentProfileManager
import com.github.catatafishen.agentbridge.ui.ThemeColor
import com.intellij.openapi.options.BoundConfigurable
import com.intellij.openapi.options.SearchableConfigurable
import com.intellij.openapi.project.Project
import com.intellij.ui.components.JBLabel
import com.intellij.ui.components.JBTextArea
import com.intellij.ui.components.JBTextField
import com.intellij.ui.dsl.builder.AlignX
import com.intellij.ui.dsl.builder.AlignY
import com.intellij.ui.dsl.builder.RowLayout
import com.intellij.ui.dsl.builder.bindItem
import com.intellij.ui.dsl.builder.bindText
import com.intellij.ui.dsl.builder.panel
import com.intellij.util.ui.JBUI
import com.intellij.util.ui.UIUtil
import com.intellij.ui.components.JBScrollPane
import java.awt.Font

@Suppress("unused")
class ClaudeCliClientConfigurable(@Suppress("UNUSED_PARAMETER") project: Project) :
    BoundConfigurable("Claude CLI"),
    SearchableConfigurable {

    private val customModelsArea = JBTextArea(8, 40).apply {
        lineWrap = false
        font = Font(Font.MONOSPACED, Font.PLAIN, JBUI.Fonts.label().size)
    }

    override fun getId(): String = ID

    override fun createPanel() = panel {
        row {
            val authNote = JBLabel(
                "<html>Run <code>claude /login</code> in a terminal to authenticate. " +
                    "Authentication problems are reported by Claude itself when you send a prompt.</html>"
            )
            authNote.foreground = UIUtil.getContextHelpForeground()
            cell(authNote)
        }
        separator()
        row("Claude binary:") {
            textField()
                .align(AlignX.FILL)
                .resizableColumn()
                .applyToComponent { (this as JBTextField).emptyText.text = "Auto-detect (leave empty)" }
                .comment("Leave empty to auto-detect on PATH.")
                .bindText(
                    { profile()?.customBinaryPath.orEmpty() },
                    { profile()?.customBinaryPath = it.trim() }
                )
        }
        row("Instructions file:") {
            textField()
                .align(AlignX.FILL)
                .resizableColumn()
                .applyToComponent { (this as JBTextField).emptyText.text = "E.g. CLAUDE.md" }
                .comment(
                    "Plugin instructions are prepended here on session start (relative to project root)."
                )
                .bindText(
                    { profile()?.prependInstructionsTo.orEmpty() },
                    { profile()?.prependInstructionsTo = it.trim() }
                )
        }
        row("Bubble color:") {
            cell(ThemeColorComboBox())
                .comment(
                    "Choose a theme-aware accent color for Claude message bubbles. " +
                        "Shared with Claude Code."
                )
                .bindItem(
                    { ThemeColor.fromKey(AcpClient.loadAgentBubbleColorKey(BUBBLE_CLIENT_TYPE)) },
                    { AcpClient.saveAgentBubbleColorKey(BUBBLE_CLIENT_TYPE, it?.name) }
                )
        }
        separator()
        row {
            label("Custom models (one per line):")
                .comment(
                    "Format: <code>&lt;model-id&gt;=&lt;Display Name&gt;</code>. " +
                        "Leave empty to use the built-in model list."
                )
        }
        row {
            cell(JBScrollPane(customModelsArea))
                .align(AlignX.FILL)
                .align(AlignY.FILL)
                .resizableColumn()
                .onIsModified { parseModels() != (profile()?.customCliModels ?: emptyList()) }
                .onApply { profile()?.customCliModels = parseModels() }
                .onReset {
                    customModelsArea.text =
                        (profile()?.customCliModels ?: emptyList()).joinToString("\n")
                    customModelsArea.caretPosition = 0
                }
        }.layout(RowLayout.PARENT_GRID).resizableRow()
    }

    private fun profile() = AgentProfileManager.getInstance()
        .getProfile(AgentProfileManager.CLAUDE_CLI_PROFILE_ID)

    private fun parseModels(): List<String> =
        customModelsArea.text.split("\n")
            .map { it.trim() }
            .filter { it.isNotEmpty() }

    companion object {
        const val ID = "com.github.catatafishen.agentbridge.client.claudecli"

        /** CSS client type shared with Claude Code (AnthropicDirect) for bubble color. */
        private const val BUBBLE_CLIENT_TYPE = "claude"
    }
}
